metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_lineage_v2.py

```diff
@@ -109,19 +109,31 @@ def get_workunits(
             f"Upstream lineage detected for {self.report.num_tables_with_upstreams} tables.",
         )
 
-        yield from self.get_table_external_upstream_workunits()
+        if self._external_lineage_map:  # Some external lineage is yet to be emitted
+            yield from self.get_table_external_upstream_workunits(discovered_tables)
+            logger.info(
+                f"Only upstream external lineage detected for {self.report.num_tables_with_external_upstreams_only} tables.",
+            )
 
-    def get_table_external_upstream_workunits(self) -> Iterable[MetadataWorkUnit]:
-        if self._external_lineage_map:
-            for (
-                dataset_name,
-                external_lineage,
-            ) in self._external_lineage_map.items():
-                upstreams = self.get_external_upstreams(external_lineage)
-                if upstreams:
-                    yield self._create_upstream_lineage_workunit(
-                        dataset_name, upstreams
-                    )
+    def get_table_external_upstream_workunits(
+        self, discovered_tables: List[str]
+    ) -> Iterable[MetadataWorkUnit]:
+        for (
+            dataset_name,
+            external_lineage,
+        ) in self._external_lineage_map.items():
+            if (
+                dataset_name not in discovered_tables
+                or not self._is_dataset_pattern_allowed(
+                    dataset_name,
+                    SnowflakeObjectDomain.TABLE,
+                )
```

Comment:
I think this is covered already:
```
    def _process_external_lineage_result_row(self, db_row):
        # key is the down-stream table name
        key: str = self.get_dataset_identifier_from_qualified_name(
            db_row["DOWNSTREAM_TABLE_NAME"]
        )
        if not self._is_dataset_pattern_allowed(key, SnowflakeObjectDomain.TABLE):
            return
```

```
    def _populate_external_lineage_from_show_query(self):
        external_tables_query: str = SnowflakeQuery.show_external_tables()
        try:
            for db_row in self.query(external_tables_query):
                key = self.get_dataset_identifier(
                    db_row["name"], db_row["schema_name"], db_row["database_name"]
                )

                if not self._is_dataset_pattern_allowed(
                    key, SnowflakeObjectDomain.TABLE
                ):
                    continue
```

To limit memory use, it seems better where it already is. If anything, we could do the `discovered_tables` check at the above 2 locations rather than here

Reply:
done.